integration/tls_to_backends_test.go

```diff
@@ -81,8 +81,13 @@ var _ = Describe("TLS to backends", func() {
 			assertWebsocketSuccess(wsApp)
 		})
 
-		It("closes connections with backends that respond with non 101-status code", func() {
-			wsApp := test.NewHangingWebSocketApp([]route.Uri{"ws-app." + test_util.LocalhostDNS}, testState.cfg.Port, testState.mbusClient, "")
+		// this test mandates RFC 6455 - https://datatracker.ietf.org/doc/html/rfc6455#section-4
+		// where it is stated that:
+		// "(...) If the status code received from the server is not 101, the
+		//       client handles the response per HTTP [RFC2616] procedures."
+		// Which means the proxy must treat non-101 responses as regular HTTP [ and not close the connection per se ]
+		It("does not close connections with backends that respond with non 101-status code", func() {
```

Comment:
So, previously, gorouter behaved as such (RFC noncompliant):
- client requests a websocket upgrade
- server responds with a  status code indicating that no upgrade will happen
- gorouter closes the client connection immediately
- the client would need to create a new connection to make a subsequent request
- the client would not see any body data returned from gorouter (but headers would have been sent)

and will now behave like this (RFC compliant):
- client requests a websocket upgrade
- server responds with a  status code indicating that no upgrade will happen
- the client receives headers + body of the response
- the connection is kept-alive for re-use

Is that accurate?

Reply:
Yes, that is correct. The main reason for this behavior is potential authentication requests by server indicated by a 401 and possible redirects via 302 / 301 to another endpoint that actually does the upgrade.